packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/PropertiesGrouping.ts

```diff
@@ -0,0 +1,323 @@
+      const mapKeyRanges = getRangesAsString(propertyGroup.ranges);
+      const lastKey = previousPropertiesInfo.length > 0 ? previousPropertiesInfo[previousPropertiesInfo.length - 1].propertyGroupKey : "";
+      const propertyGroupKey = `${lastKey}:${propertyGroup.propertyName}(${mapKeyRanges})`;
+      const mapKey = `${byProperties.fullClassName}:${propertyGroupKey}`;
+
+      if (!uniqueProperties.get(mapKey)) {
+        uniqueProperties.set(mapKey, {
+          ecClass: await getClass(metadata, byProperties.fullClassName),
+          propertyGroup: {
+            propertyName: propertyGroup.propertyName,
+            ranges: propertyGroup.ranges,
+          },
+          previousPropertiesGroupingInfo: previousPropertiesInfo.map((groupingInfo) => ({
+            fullClassName: byProperties.fullClassName,
+            propertyGroup: groupingInfo.propertyGroup,
+          })),
+        });
+      }
+
+      previousPropertiesInfo.push({
+        propertyGroup,
+        propertyGroupKey,
+      });
+    }
+  }
+  // Order might change in uniqueProperties, resorting to make sure that properties with fewer previous properties are returned first.
+  return [...uniqueProperties.values()].sort((lhs, rhs) => lhs.previousPropertiesGroupingInfo.length - rhs.previousPropertiesGroupingInfo.length);
+}
+
+function getRangesAsString(ranges?: Range[]): string {
+  return ranges
+    ? ranges
+        .map((range) => `${range.fromValue}-${range.toValue}(${range.rangeLabel ? `${range.rangeLabel}` : ""})`)
+        .sort()
+        .join(";")
+    : "";
+}
+
+async function shouldCreatePropertyGroup(
+  metadata: IMetadataProvider,
+  extractedPropertyInfo: PropertyGroupInfo,
+  nodePropertyGroupingParams: HierarchyNodePropertiesGroupingParams,
+  nodeFullClassName: string,
+): Promise<boolean> {
+  if (
+    nodePropertyGroupingParams.fullClassName !== extractedPropertyInfo.ecClass.fullName ||
+    nodePropertyGroupingParams.propertyGroups.length < extractedPropertyInfo.previousPropertiesGroupingInfo.length + 1
+  ) {
+    return false;
+  }
+  const currentProperty = nodePropertyGroupingParams.propertyGroups[extractedPropertyInfo.previousPropertiesGroupingInfo.length];
+  if (
+    currentProperty.propertyName !== extractedPropertyInfo.propertyGroup.propertyName ||
+    !doRangesMatch(currentProperty.ranges, extractedPropertyInfo.propertyGroup.ranges)
+  ) {
+    return false;
+  }
+  if (!doPreviousPropertiesMatch(extractedPropertyInfo.previousPropertiesGroupingInfo, nodePropertyGroupingParams)) {
+    return false;
+  }
+  const nodeClass = await getClass(metadata, nodeFullClassName);
+  if (!(await nodeClass.is(extractedPropertyInfo.ecClass))) {
+    return false;
+  }
+  return true;
+}
+
+/** @internal */
+export function doPreviousPropertiesMatch(
+  previousPropertiesGroupingInfo: PreviousPropertiesGroupingInfo,
+  nodesProperties: HierarchyNodePropertiesGroupingParams,
+): boolean {
+  return previousPropertiesGroupingInfo.every(
```

Comment:
`doPreviousPropertiesMatch` is called only in one place. `shouldCreatePropertyGroup` calls it after checking if 
`nodePropertyGroupingParams.propertyGroups.length < handlerGroupingParams.previousPropertiesGroupingInfo.length + 1`.
It could throw if it were to be called somewhere else. 
But another note: `nodesProperties.propertyGroups` can't be the same length as `previousPropertiesGroupingInfo`. It has to be at larger by at least one item.

Reply:
This is an exported function, which makes it available for other modules to import. Please at least add a function comment explaining the input constraints.

Reply:
Added a check to the function:
`previousPropertiesGroupingInfo.length <= nodesProperties.propertyGroups.length && <rest of the code>`